Update Networking layer with GET `/wc-analytics/customers?search`

iamgabrielma opened the pull request:
Part of #7741 

### Description

For merchants to be able to see and choose existing customers from their store when creating or editing orders, we need to know who are these existing customers. At the moment we can only get partial customer data based on other API properties, like Orders.

### Scope

This PR adds the initial networking layer that we need in order to fetch existing customers from a store via a GET request to the`/wp-json/wc-analytics/customers?search=` API endpoint

### Changes
- Adds `WCAnalyticsCustomerRemote`, `WCAnalyticsCustomerMapper`, and `WCAnalyticsCustomer` entities to the Networking layer. This retrieves, decodes, maps, and models a `WCAnalyticsCustomer` object from its JSON representation.
- Adds a `wc-analytics-customers.json` file, replicating a `/wp-json/wc-analytics/customers?search=` API response to use through `MockNetwork`.

### Testing instructions
- Make sure that Unit Tests pass

iamgabrielma committed:
Add initial WCAnalyticsCustomer model, mapper, remote

iamgabrielma committed:
Change Remote method to return a collection. Add tests.

iamgabrielma committed:
Add retrieveCustomers method to Remote

iamgabrielma committed:
Add WCAnalyticsCustomerMapper Tests

iamgabrielma committed:
Add map with search value for WCAnalyticCustomer

iamgabrielma committed:
Add unit tests for retrieveCustomerByName

iamgabrielma commented on Networking/Networking/Remote/WCAnalyticsCustomerRemote.swift:
```diff
@@ -0,0 +1,45 @@
+    ///     - name: Name of the customer that will be retrieved
+    ///     - completion: Closure to be executed upon completion.
+    ///
+    func retrieveCustomerByName(for siteID: Int64, with name: String, completion: @escaping (Result<[WCAnalyticsCustomer], Error>) -> Void) {
```
I'm assuming we may have more than one customer with the same name ( specially once we add support for partial search, we'll get more results than one match ), that's why for now I'm returning an `[WCAnalyticsCustomer]` entity here.

joshheald commented on Networking/Networking/Mapper/WCAnalyticsCustomerMapper.swift:
```diff
@@ -0,0 +1,33 @@
+    /// (Attempts) to convert a dictionary into a `WCAnalyticsCustomer` entity
+    ///
+    func mapUniqueCustomer(response: Data, searchValue: String? = "") throws -> WCAnalyticsCustomer? {
+        if searchValue == "" {
+            return WCAnalyticsCustomer(userID: 0, name: "")
+        }
```
This... seems strange. Why not return nil, which you'd get for free by deleting this code and leaving it to the filter to deal with it? If you return this, it'll go into storage or you'll have to code around it in the rest of the app.

iamgabrielma replied:
You're totally right, I'm actually not sure why I did that. Fixed on 527f5f9319a299ce5b8f215d56a5c44a6cac991a (but later on, then has been entirely removed on def5158743fcae7d44aeef8c931e23a788d77860 😬 )

joshheald commented on Networking/Networking/Model/WCAnalyticsCustomer.swift:
```diff
@@ -0,0 +1,35 @@
+
+extension WCAnalyticsCustomer {
+    enum CodingKeys: String, CodingKey {
+        case userID =   "id"
+        case name   =   "username"
```
Have you checked these against the Android implementation?

iamgabrielma replied:
Yeah! Android did not follow this approach initially, and it was decided to launch the feature as-it-is ( p1656431752334189-slack-CGPNUU63E ) and iterate from there based on feedback: Either leave it as it is,  implementing the proposed solution we're doing now on iOS ( pe5pgL-3r-p2 ), or using the API directly if a fix for searching first and last names on the `wc/v3/customers` endpoint came up ( it hasn't at the moment ).

So in this case I'm going a bit blind on this regard, but I asked for current implementation details ( p1664485668537849-slack-CGPNUU63E ) so I could follow their approach as well if needed. At this point I believe I will only need the `userID` in this case, but left the `username` for the moment so I could play with different datatypes if needed.

iamgabrielma replied:
Adding to this: Just wondering if would be helpful to add some extra fields here that (despite being optional) could be helpful to store the initial information for each customers, and not only their customer ID and username. For example this endpoint does not separate between firstName and lastName, as we want it to appear in the end, however there are other fields that do coincide, for example: Country, City, or PostCode. 

On one end I understand that we're already performing the request to fetch this data we could fill all possible fields in one go, however on other end, it may cause the problems we spoke in our last pairing when attempting to map and decode the data, as we could find unexpected input.

For now I'm inclined to use as little data as needed for this step (that's basically the customerID) and add the rest directly on the next steps when we actually fetch data from the specific `/wc/v3/customers/<id>` endpoint, which contains all we need. Thoughts?

joshheald replied:
> For now I'm inclined to use as little data as needed for this step (that's basically the customerID) and add the rest directly on the next steps when we actually fetch data from the specific /wc/v3/customers/<id> endpoint, which contains all we need. Thoughts?

That's fine with me. You can save yourself some hassle if you get the overall set of properties correct when you do the Storage layer, so that you can make a single change to the Core Data model, but iterating on the Networking layer later is not expensive.

joshheald commented on Networking/NetworkingTests/Remote/WCAnalyticsCustomerRemoteTests.swift:
```diff
@@ -0,0 +1,112 @@
+        XCTAssert(result.isFailure)
+    }
+
+    func test_WCAnalyticsCustomerRemote_when_calls_retrieveCustomers_then_returns_all_parsed_customers_successfully() throws {
+        // Given
+        network.simulateResponse(requestUrlSuffix: "customers", filename: "wc-analytics-customers")
+
+        // When
+        let result = waitFor { promise in
+            self.remote.retrieveCustomers(for: self.sampleSiteID) { result in
+                promise(result)
+            }
+        }
+        let customers = try XCTUnwrap(result.get())
+
+        // Then
+        XCTAssertNotNil(customers)
+        XCTAssertEqual(customers[0].userID, 1)
+        XCTAssertEqual(customers[1].userID, 2)
+        XCTAssertEqual(customers[0].name, "John")
+        XCTAssertEqual(customers[1].name, "Paul")
+    }
```
It's probably worth doing this test for customer search too. Once you add that, you can delete both the `isSuccess` tests, as they're just simpler versions of the same. 

Checking that the customers are actually returned, not just a success result (which may be empty) is valuable, as returning the mapped data is part of the Remote's responsibilities.

You don't necessarily need to check both properties here, as that is tested on the Mapper. Maybe just that the names are as you expect, or that there are two customers in the returned data, but still more than `isSuccess`.

iamgabrielma replied:
Thanks!

> Checking that the customers are actually returned, not just a success result (which may be empty) is valuable, as returning the mapped data is part of the Remote's responsibilities.

Following your advice here and checking what this actually returns, this has uncovered a problem with the way `retrieveCustomerByName()` worked, and that's was always retuning the full array of entities, not the filtered-by-name entities. 

I'm assuming this is because our Mapper class just calls `map()`, not other methods from the subclass implementation details. In order to solve this, I moved the filtering logic to the Remote instead: https://github.com/woocommerce/woocommerce-ios/pull/7780/commits/def5158743fcae7d44aeef8c931e23a788d77860 . I'm not positive this is a good way to do it (specially the usage of `.contains()` and `.filter()` ), but seems to resonate better with other cases I've checked in our codebase. Another alternative could have been to create a new mapper entirely for this, but seemed a bit overkill. Happy to go through it together, or iterate on it from any feedback 👍

iamgabrielma replied:
Forgot to add: At this point I can't be sure yet how it will work, but my guess is that once all customers are fetched, we could also use [FetchedResults](https://developer.apple.com/documentation/swiftui/fetchedresults) in the Storage layer to request a particular set of results from local storage, rather than having to perform further requests through the network.

joshheald replied:
> I'm assuming this is because our Mapper class just calls map(), not other methods from the subclass implementation details. In order to solve this, I moved the filtering logic to the Remote instead: https://github.com/woocommerce/woocommerce-ios/commit/def5158743fcae7d44aeef8c931e23a788d77860 . I'm not positive this is a good way to do it (specially the usage of .contains() and .filter() ), but seems to resonate better with other cases I've checked in our codebase. Another alternative could have been to create a new mapper entirely for this, but seemed a bit overkill.

Gut feeling without having looked at any of the code: the Networking layer should probably just return what the API does, it shouldn't go beyond correctly mapping entities and making requests and such. The filtering sounds more like business logic.

If the API doesn't give us filtered results back, that business logic should be handled in Yosemite, and as you suggest in your other comment, this may well mean leveraging CoreData's filtering to do it.

iamgabrielma replied:
> Gut feeling without having looked at any of the code: the Networking layer should probably just return what the API does, it shouldn't go beyond correctly mapping entities and making requests and such. The filtering sounds more like business logic.
> 
> If the API doesn't give us filtered results back, that business logic should be handled in Yosemite, and as you suggest in your other comment, this may well mean leveraging CoreData's filtering to do it.

Thanks, that makes a lot of sense. Most of my doubts regarding this comes from trying to understand how Networking and Core Data will work together, but I'm getting a clearer idea as we iterate 👍

joshheald reviewed (approved):
👍 A few suggestions but generally looks good.

Most important one is the question of what you do when there's no search term – I may be missing something here, so say if you want to discuss that case a little more.

iamgabrielma committed:
Return nil when no searchValue is given

iamgabrielma committed:
Remove mapUniqueCustomer in favor of its Remote

When Unit Testing the previous retrieveCustomerByName(), we uncovered an error where the resultant array of entities was not only the used search term, but all entities from the endpoint. This is because the Mapper calls only the WCAnalyticsCustomerMapper.map() function, and not other methods. In this case, we remove the mapUniqueCustomer() from the Mapper, and add the logic to WCAnalyticsCustomerRemote instead.

iamgabrielma committed:
Merge branch 'trunk' into issue/7741-update-networking-analytics

iamgabrielma committed:
Modify WCAnalyticsCustomer decoding CodingKeys

iamgabrielma committed:
Remove retrieveCustomers() from Remote. Fix tests.

iamgabrielma commented:
Thanks for your time and the walkthrough during the pairing session @joshheald, following what we checked yesterday and how we're not going to store the whole API response, I made some modifications here:

- Modified the `CodingKeys`, we no longer attempt to decode `id` and `username`, but `user_id` and `name`.
- Removed the `retrieveCustomers()` remote method and left `retrieveCustomersByName()`, so we can fetch specifically those which contains the specified keyword.
- Adapted the Unit tests

iamgabrielma committed:
Add data CodingKeys to mapper decodable

iamgabrielma committed:
Add query parameters to searchCustomers Unit Tests
